Implement resource based permissions control.

ycheng-aa opened the pull request:
1. Provide 3 APIs, resource-permissions,
   current-user-resource-permissions and
   group-resource-permissions.
2. Resources are APIs which need to be included in permission
   control. Permissions are creat, read, update, delete. PDC
   will collect resource permissions accroding to router
   automatically.
3. resource-permissions API provide list function to
   list all available resource permissions.
   current-user-resource-permissions list current
   user's resource permissions.
4. Through group-resource-permissions, resource permission
   could be assigned to or removed from a group. Users in
   that group will be affected.
5. Prevent non-superuser from accessing admin page.

JIRA: PDC-1384

lubomir commented on pdc/apps/auth/migrations/0004_auto_20160512_1051.py:
```diff
@@ -0,0 +1,56 @@
+
+
+def create_action(apps, schema_editor):
+    """Migrate contacts stored directly on release components."""
```
This comment does not seem to match the method.

ycheng-aa replied:
Fixed in 71b3211d41ee6bd4

lubomir commented on pdc/apps/auth/views.py:
```diff
@@ -527,3 +530,134 @@ def list(self, request):
+
+class GroupResourcePermissionViewSet(common_viewsets.PDCModelViewSet):
+    """
+    This end-point provides t group resource permissions.
```
Stray *t*.

ycheng-aa replied:
Fixed in 71b3211d41ee6bd4

lubomir commented on pdc/apps/auth/views.py:
```diff
@@ -527,3 +530,134 @@ def list(self, request):
+
+        __Response__:
+
+
```
Is response description missing here?

ycheng-aa replied:
Fixed in 71b3211d41ee6bd4

lubomir commented on pdc/apps/utils/middleware.py:
```diff
@@ -27,3 +34,50 @@ def process_response(self, request, response):
+    def process_request(self, request):
+        if request.user.is_authenticated() and request.path.startswith(reverse('admin:index')):
+            if not (request.user.is_active and request.user.is_superuser):
+                return HttpResponseRedirect("/%s%s/" % (settings.REST_API_URL, settings.REST_API_VERSION))
```
Why is this needed? The behaviour up to now was to allow login to admin section for users who are marked as *staff*.

ycheng-aa replied:
I thought if we grant all model permissions to all groups, then API permissions could work all by itself. PDC will not have the situation that it grant a group a API's permission, but still can't work because lack of model permission. But it will bring the risk that user could manipulate model directly to avoid permission control.

I think you mentioned we should remove model permissions. But I am not sure what it means.  Did you mean remove `django.contrib.auth.backends.ModelBackend` from `AUTHENTICATION_BACKENDS`? I tested to remove it and it looks we can't login admin page.

lubomir commented on pdc/apps/auth/views.py:
```diff
@@ -527,3 +530,134 @@ def list(self, request):
+        __URL__: $LINK:currentuserpermissions-list$
+
+        __Response__:
+            [
```
This does not render correctly. Maybe a blank line is missing?

ycheng-aa replied:
Fixed in 71b3211d41ee6bd4

lubomir commented:
The code looks reasonable to me. I'm not a big fan of adding another dependency, but if it makes things easier, then why not.

I think it would be nice to have the new urls grouped under the `auth/` prefix. How about removing the `resource` keyword from the URL too? Does it add any value?

A couple notes from my testing:
 * Running a development server with this patch suddenly requires login to access any API. How can the server be configured to not require any authentication for read access? This is a requirement for deploying in Fedora infrastructure.
 * Even though I'm an admin, visiting `/current-user-resource-permission` does not show anything. Furthermore, why add another end-point here instead of merging this into existing `/auth/current-user`?
 * Urls in documentation for new APIs don't show correctly.
 * Group resource permissions endpoint does not have *Create* tab, but it has *Bulk Create*.

ycheng-aa commented:
@lubomir  Thanks you very much for the careful review.

* I moved new endpoints under `auth/`. Merged resource permission content to `/auth/current-user`.
* For superuser, It will show all resource permissions even not explicitly assigned to him and doesn't   exist in DB.
* Fixed documentation and make sure `create` tab work.
* In settings.py, add a flag to allow anyone has read permission.

lubomir commented:
Regarding the model permissions, I think it does not matter if they are used or not as long as they are not exposed to users (including admins). It seems confusing to have to different permission systems. What is the expected workflow?

ycheng-aa commented:
@lubomir I am hoping we could grant all model permissions to all groups. Then use resource based permissions control only.  This is my expected workflow.
Because user could have all model permissions, grant any resource permission to a user could make sure user really has the permission. Normal user also couldn't modify model directly instead of by accessing API. 
All this just my conjecture, I am not sure if there is something wrong in the workflow.

lubomir commented:
That sounds reasonable to me. If everyone has all the model permissions, are changes in the admin section still limited to admins only?

lubomir commented:
Anyway, my proposal here would be to merge this pull request and deploy it on devel server. Then we can test how it behaves, talk to users and have them try it. We have quite some time until next release to fix any issues.

ycheng-aa commented:
Changes you mean the page like https://pdc.host.dev.eng.pek2.redhat.com/changes/ ? If it's this page, it's not in admin section.
I am also not sure if the admin restriction is necessary. As you said, we could deploy then test and fix problems. Thanks.

simozhan commented:
Looks good to me. :+1: Agree with @lubomir, merge it first and test it.
